### ISSUE TITLE
Fit parameters order of magnitude different

### ISSUE THREAD

**Issue by AlexisRenchon** [JuliaNLSolvers/LsqFit.jl#200]
Hi, 

I am using a (soil respiration) model that predict soil respiration as a function of soil temperature (input 1) and soil moisture (input 2). Below is a link to the model: 

https://github.com/AlexisRenchon/DAMMmodel.jl/blob/master/src/DAMM.jl

my problem, which can occur with any model, is that the parameters have order of magnitude differences, for example my initial param are

Param_ini = [1e8, 59, 1e-6, 2.0e-3, 0.4, 0.0125]

you can see the range of magnitude, from e-6 to e8. 

I know that this is often a problem for fitting algorithm, and a solution if to add the factors (e.g. *e8 or *e-6) in the model, so that the param to fit are of similar order of magnitude. 

Is this the only solution? I find it not ideal to add factors in the model.

**Comment by AndiMD** [comment]
I came across a similar issue #178. How about an option to prescale/transform parameters? This could generally be useful to improve ill-conditioned problems, e.g.
```curve_fit(..., param_prescale=(1E8, log, 1E3) )```
or more generally
```curve_fit(..., param_prescale=(x->1E-8x, x->log(x), x->1E3*x), param_backscale=(x->1E8x, x->exp(x), x->1E-3*x) )```

**Comment by AlexisRenchon** [comment]
Hi AndiMD,
What you suggest is exactly what I am looking for. It would be perfect, and would help many I am sure. 
If I understand correctly, it doesn't exist yet but you are suggesting to implement it? 

In the meantime, I probably have to create another temporary script for my model package with those scaling factors, which is not very elegant... I will look forward to this option being implemented in LsqFit.jl

**Comment by AndiMD** [comment]
Hi Alexis,
I am just a user with a similar issue. Whether my suggestion is acceptable for the LsqFit package must be decided by the package maintainers.

**Comment by AndiMD** [comment]
In the meantime, this issue can be hacked-around like so (untested):
```using LsqFit
import LsqFit.curve_fit
import ForwardDiff
"""
curve_fit where the model(x,p) function is transformed by with prescaled parameters p = scale_fwd(scale_bck(p)):
    model_scaled(x,p) = model(x, scale_bck(p))
    model_scaled(x,scale_fwd(p)) == model(x,p)
"""
function curve_fit(model, xdata::AbstractArray, ydata::AbstractArray, p0::AbstractArray, scale_fwd::Function, scale_bck::Function; inplace, kwargs...)
    
    model_scaled(x, p_fwd) = model(x, scale_bck(p_fwd))
    p0_fwd = scale_fwd(p0)
    res_fwd = curve_fit(model_scaled, xdata, ydata, p0_fwd; inplace=inplace, kwargs...)
    
    param_reconstructed = scale_bck(res_fwd.param)
    jac_fwd_inv = ForwardDiff.jacobian(scale_bck, res_fwd.param) |> inv
    resid_reconstructed = model.(xdata, (param_reconstructed,) ) .- ydata
    
    res_reconstructed = LsqFit.LsqFitResult(param_reconstructed, resid_reconstructed, res_fwd.jacobian*jac_fwd_inv, res_fwd.converged, res_fwd.wt)
end

model(x,p) = @. p[1]*exp(p[2]*x)
xdata = range(0.1,3,length=5) # fake data x
ydata = model(xdata,[0.5,2.0]).+0.1*rand(5) # fake data y
p0 = [1.0,1.0]
res1 = curve_fit(model, xdata, ydata, p0, p->[1E-3*p[1], exp(p[2])], p->[1E3*p[1], log(p[2])], inplace=false, g_tol=1E-16, x_tol=1e-16)

```

**Comment by AlexisRenchon** [comment]
Thanks AndiMD, I am going to test your solution today or tomorrow and keep you posted

**Comment by AlexisRenchon** [comment]
This was solved by the same solution as #201 
e.g., (not a MWE) 
```
function fitDAMM(Ind_var, Resp)
  lb = [0.0, 0.0, 0.0, 0.0] # params can't be negative
  p_ini = [0.3, 6.4, 1.0, 1.0] # initial parameters
  p_fact = (1e9, 1e1, 1e-6, 1e-4, 1, 1) # factor of param, better fit
  DAMMfix = (x,p) -> DAMM(x, p_fact.*(p[1], p[2], p[3], p[4], poro_val, 0.02))
  fit_bounds = curve_fit(DAMMfix, Ind_var, Resp, p_ini, lower=lb)
  p = fit_bounds.param # fitted param result 
  fparam = p_fact[1:4].*p # multiply them by p_fact
  params = [fparam[1], fparam[2], fparam[3], fparam[4], poro_val, 0.02]
  return params
end
```
